include/ignition/math/PiecewiseScalarField3.hh

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef IGNITION_MATH_PIECEWISE_SCALAR_FIELD3_HH_
+#define IGNITION_MATH_PIECEWISE_SCALAR_FIELD3_HH_
+
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <utility>
+#include <vector>
+
+#include <ignition/math/Region3.hh>
+#include <ignition/math/Vector3.hh>
+#include <ignition/math/config.hh>
+
+namespace ignition
+{
+  namespace math
+  {
+    // Inline bracket to help doxygen filtering.
+    inline namespace IGNITION_MATH_VERSION_NAMESPACE {
+    //
+    /** \class PiecewiseScalarField3 PiecewiseScalarField3.hh\
+     * ignition/math/PiecewiseScalarField3.hh
+     */
+    /// \brief The PiecewiseScalarField3 class constructs a scalar field F
+    /// in R^3 as a union of scalar fields Pn, defined over regions Rn i.e.
+    /// piecewise.
+    ///
+    /// \tparam ScalarField3T a callable type taking a single Vector3<ScalarT>
+    ///   value as argument and returning a ScalarT value. Additionally:
+    ///   - for PiecewiseScalarField3 to have a stream operator overload,
+    ///     ScalarField3T must support stream operator overload;
+    ///   - for PiecewiseScalarField3::Minimum to be callable, ScalarField3T
+    ///     must implement a
+    ///     ScalarT Minimum(const Region3<ScalarT> &, Vector3<ScalarT> &)
+    ///     method that computes its minimum in the given region and returns
+    ///     an argument value that yields said minimum.
+    /// \tparam ScalarT a numeric type for which std::numeric_limits<> traits
+    ///   have been specialized.
+    ///
+    /// ## Example
+    ///
+    /// \snippet examples/piecewise_scalar_field3_example.cc complete
+    template<typename ScalarField3T, typename ScalarT>
+    class PiecewiseScalarField3
+    {
+      /// \brief A scalar field P in R^3 and
+      /// the region R in which it is defined
+      public: struct Piece {
+        Region3<ScalarT> region;
+        ScalarField3T field;
+      };
+
+      /// \brief Constructor
+      public: PiecewiseScalarField3() = default;
+
+      /// \brief Constructor
+      /// \param[in] _pieces scalar fields Pn and the regions Rn
+      ///   in which these are defined. Regions may not overlap.
```

Comment:
When I first read this I understood that regions may or may not overlap. But seeing the error message below, I guess this should be:

```suggestion
      ///   in which these are defined. Regions should not overlap.
```

?

Alternatively, it looks like there's a resolution if the regions do overlap. So we could remove the "Detected overlap..." error message and document that

```suggestion
      ///   in which these are defined. When regions overlap, the first one takes precedence.
```

What do you think?

Reply:
> Alternatively, it looks like there's a resolution if the regions do overlap.

@chapulina I would've thrown an exception, but it seems you guys have done an effort in keeping `ignition-math` mostly exception-safe.

The issue with establishing a precedence is that it creates a discrepancy between definition and behavior. You'll inspect or print the scalar field and you'll see cases for when, say, `x < 1` and `x > 1` that behave as if you had set them to `x < 1` and `x >= 1`.

Reply:
Rephrased documentation in 38cbc349121a83d2b94df6328f581fa38aeb127e